### PR TITLE
Convert user extensions to dictionaries

### PR DESCRIPTION
**Pull request by jjcasmar**
Types used by user extensions are unknown and when the gltf dictionary is created, it can produce dictionaries with non serializable data. 

Recursively descend into extensions to convert non serializable data into a dictionary.

This allows for example to add a `TextureInfo` to an extension and export correctly, which is needed for example for pbrSpecularGlossiness

I want to thanks @scurest for helping with this issue. 
https://github.com/KhronosGroup/glTF-Blender-IO/issues/845

### CODE REVIEW

**Comment by jjcasmar** [comment]
@julienduroure @emackey

**Comment by emackey** [comment]
@scurest The conversation in #845 is a little hard to follow.  Do you think this is robust?

File: addons/io_scene_gltf2/io/com/gltf2_io.py [hunk 1]
```diff
@@ -117,7 +129,7 @@ def to_dict(self):
         result["bufferView"] = from_int(self.buffer_view)
         result["byteOffset"] = from_union([from_int, from_none], self.byte_offset)
         result["componentType"] = from_int(self.component_type)
-        result["extensions"] = from_union([lambda x: from_dict(lambda x: from_dict(lambda x: x, x), x), from_none],
+        result["extensions"] = from_union([lambda x: from_dict(extension_to_dict, x), from_none],
```
Comment (scurest) [inline addons/io_scene_gltf2/io/com/gltf2_io.py]: I like this better than what I did since it keeps it more consistent.

But it's technically now possible to do

    "extensions" : {"my_great_extension": 42}   // ie. extension value not a dict

whereas before this was impossible. (That's what I had two functions for in the PoC.) Could you replace it with a call to, like,

````python
def from_extension(x):
    x = extension_to_dict(x)
    assert isinstance(x, dict)
    return x
````

File: addons/io_scene_gltf2/io/com/gltf2_io.py [hunk 2]
```diff
@@ -1170,12 +1182,12 @@ def to_dict(self):
...
                                            self.buffer_views)
         result["cameras"] = from_union([lambda x: from_list(lambda x: to_class(Camera, x), x), from_none], self.cameras)
-        result["extensions"] = from_union([lambda x: from_dict(lambda x: from_dict(lambda x: x, x), x), from_none],
+        result["extensions"] = from_union([lambda x: from_dict(extension_to_dict, x), from_none],
                                           self.extensions)
         result["extensionsRequired"] = from_union([lambda x: from_list(from_str, x), from_none],
                                                   self.extensions_required)
         result["extensionsUsed"] = from_union([lambda x: from_list(from_str, x), from_none], self.extensions_used)
-        result["extras"] = self.extras
+        result["extras"] = extension_to_dict(self.extras)
```
Comment (scurest) [inline addons/io_scene_gltf2/io/com/gltf2_io.py]: I think you missed this one.

**Comment by scurest** [comment]
@emackey  I'll try to summarize the issue. Here's how things normally work

* A TextureInfo stores a Texture object in it's `texture` property during the gathering phase
* After the gathering phase there is a __traverse pass where the Texture is moved into the top-level `gltf.textures` array and the `texture` property is replaced with an index into that array.
* After that there is a to_dict pass where all the gltf2_io classes are turned into dicts in preparation for JSON dumping. to_dict checks types and expects the `texture` property to be an integer.

For extensions the problem is

1. You can't put the TextureInfo object in your extension directly because the to_dict pass doesn't descend into extensions and turn objects into dicts like it does for normal properties.
2. You can't turn the TextureInfo object into a dict with to_dict during the gathering phase like eg. gltf2_blender_gather_lights does for lights because you can only call to_dict after the __traverse pass has replaced `texture` with an integer. (This isn't a problem for current extensions since they never use things like TextureInfo that are affected by __traverse.)

So this PR changes (1).

----

Like jjcasmar says, this is really the natural place for this to happen. This is what happens to everything else in the to_dict pass, and it's exactly where this would happen if to_dict _did_ understand the structure of extensions.

So it seems pretty safe to me.

**Comment by jjcasmar** [comment]
Do we want to support also for extras? I'm not sure if extras should be so complex.

**Comment by scurest** [comment]
I think its fine if this PR just does extensions.

**Comment by emackey** [comment]
@jjcasmar Can you make the changes mentioned in @scurest's review above?  Thanks.

**Comment by jjcasmar** [comment]
Done. I have also added the extras export.

**Comment by scurest** [comment]
@jjcasmar It doesn't seem to have changed (maybe you forgot to `git add` or something?).

**Comment by jjcasmar** [comment]
@scurest should be now

**Comment by scurest** [comment]
extras don't have to be dicts. Also you missed updating the `Gltf` class.

**Comment by emackey** [comment]
@jjcasmar The force-pushing is making it difficult to track all the changes coming into this branch.  Do you mind making the remaining changes as normal commits, without squash/force-push?

**Comment by scurest** [comment]
@emackey (You might already know this, but you can click on the words "force-pushed" in the "jjcasmar force-pushed the..." lines to see a diff.)

**Comment by jjcasmar** [comment]
I think all the comment have been addressed, right?

**Comment by scurest** [comment]
* `from_extra` requires that the extras be either a dict or None. That isn't right, the spec does not constrain the type of extras. You can just do this

    ```py
    def from_extra(x):
        return extension_to_dict(x)
    ```

* You didn't update the `Gltf` class. It is still the odd man out. It looks like this

    ```py
    result["extras"] = extension_to_dict(self.extras)
    ```

    while all the other classes look like this
    ```py
    result["extras"] = from_extra(self.extras)
    ```

**Comment by jjcasmar** [comment]
You are right, I completely missed that comments. Now they are addressed

**Comment by emackey** [comment]
@scurest Let me know if this works for you.

@julienduroure We should probably include this in 2.82 since 3rd-party extensions are new in that version.

**Comment by scurest** [comment]
@jjcasmar Cool. I tested everything. Lgtm :+1:
